src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs

```diff
@@ -525,28 +525,37 @@ private static Task GetHostEntryOrAddressesCoreAsync(string hostName, bool justR
 
                 asyncState = family == AddressFamily.Unspecified ? (object)ipAddress : new KeyValuePair<IPAddress, AddressFamily>(ipAddress, family);
             }
-            else if (NameResolutionPal.SupportsGetAddrInfoAsync)
+            else
             {
+                if (NameResolutionPal.SupportsGetAddrInfoAsync)
+                {
 #pragma warning disable CS0162 // Unreachable code detected -- SupportsGetAddrInfoAsync is a constant on *nix.
 
-                // If the OS supports it and 'hostName' is not an IP Address, resolve the name asynchronously
-                // instead of calling the synchronous version in the ThreadPool.
+                    // If the OS supports it and 'hostName' is not an IP Address, resolve the name asynchronously
+                    // instead of calling the synchronous version in the ThreadPool.
 
-                ValidateHostName(hostName);
+                    ValidateHostName(hostName);
 
-                if (NameResolutionTelemetry.Log.IsEnabled())
-                {
-                    return justAddresses
-                        ? (Task)GetAddrInfoWithTelemetryAsync<IPAddress[]>(hostName, justAddresses, family, cancellationToken)
-                        : (Task)GetAddrInfoWithTelemetryAsync<IPHostEntry>(hostName, justAddresses, family, cancellationToken);
-                }
-                else
-                {
-                    return NameResolutionPal.GetAddrInfoAsync(hostName, justAddresses, family, cancellationToken);
+                    Task t;
+                    if (NameResolutionTelemetry.Log.IsEnabled())
+                    {
+                        t = justAddresses
+                            ? (Task)GetAddrInfoWithTelemetryAsync<IPAddress[]>(hostName, justAddresses, family, cancellationToken)
+                            : (Task)GetAddrInfoWithTelemetryAsync<IPHostEntry>(hostName, justAddresses, family, cancellationToken);
+                    }
+                    else
+                    {
+                        t = NameResolutionPal.GetAddrInfoAsync(hostName, justAddresses, family, cancellationToken);
+                    }
+
+
+                    if (!t.IsFaulted || !(t.Exception?.InnerException is InvalidOperationException))
```

Comment:
I'm a learner at the task based async model, but it looks like `IsFaulted` does not wait for completion. Is the task guaranteed to have completed by the time it is returned by GetAddrInfoAsync? If not, is there a race here? 

Another option maybe is to wrap this task in another, which knows to go the other route if the inner task fails.

Reply:
I'm looking at here https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs#L1514

Reply:
I was thinking about changing the return types to for example be ValueTask but I decided not to do that for rare (IMHO) case. 
As far as the state: I use [ Task.FromException](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.fromexception?view=net-5.0 ) in the pal and that claims:
 > This method creates a Task object whose Status property is Faulted and whose Exception property contains exception.

Reply:
Ah, of course. So the task is indeed guaranteed to be completed if it has faulted immediately. It just might not be completed yet if the fault is reported in the GetAddressInfoExCallback. For this particular workaround, maybe that race doesn't matter.

Reply:
yes, any errors from the completion callback should be propagated to caller as far as I can tell. (e.g. no retry)